pinot-core/src/main/java/com/linkedin/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -367,20 +380,95 @@ private void setGroupByResults(@Nonnull BrokerResponseNative brokerResponseNativ
       }
       finalResultMaps[i] = finalResultMap;
     }
+    //If HAVING clause is set, we further filter the group by results based on the HAVING predicate
+    if (havingFilterQuery != null) {
+      HavingClauseComparisonTree havingClauseComparisonTree =
+          HavingClauseComparisonTree.buildHavingClauseComparisonTree(havingFilterQuery, havingFilterQueryMap);
+
+      //Applying close policy
+      //We just keep those groups (from different aggregation functions) that are exist in the result set of all aggregation functions.
+      //In other words, we just keep intersection of groups of different aggregation functions.
+      for (int i = 0; i < numAggregationFunctions; i++) {
+        Iterator<Map.Entry<String, Comparable>> finalMapIterator = finalResultMaps[i].entrySet().iterator();
+        while (finalMapIterator.hasNext()) {
+          Map.Entry<String, Comparable> aggResult = finalMapIterator.next();
+          for (int j = 0; j < numAggregationFunctions; j++) {
+            if (j != i) {
+              if (finalResultMaps[j].get(aggResult.getKey()) == null) {
+                finalMapIterator.remove();
+                break;
+              }
+            }
+          }
+        }
+      }
+
+      //Now it is time to remove those groups that do not validate HAVING clause predicate
+      //We first find which group keys need to be deleted and then remove them in the next step
+      Vector<String> removedGroupIDs = new Vector<String>();
+      Iterator<Map.Entry<String, Comparable>> finalMapIterator = finalResultMaps[0].entrySet().iterator();
+      Map<String, Comparable> singleGroupAggResults = new TreeMap<String, Comparable>(String.CASE_INSENSITIVE_ORDER);
+
+      while (finalMapIterator.hasNext()) {
+        int i;
+        Map.Entry<String, Comparable> aggResult = finalMapIterator.next();
+        singleGroupAggResults.put(columnNames[0], aggResult.getValue());
+        for (i = 1; i < numAggregationFunctions; i++) {
+          singleGroupAggResults.put(columnNames[i], finalResultMaps[i].get(aggResult.getKey()));
+        }
+        if (!havingClauseComparisonTree.isThisGroupPassPredicates(singleGroupAggResults)) {
+          removedGroupIDs.add(aggResult.getKey());
```

Comment:
Doesn't it make more sense to remove the group immediately from the iterator and the other aggregation groups instead of storing the keys to remove, only to remove them in a later phase?